Mds/new mobile sdk/okta 503880 revise quickstart (#3424)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/sign-into-mobile-app-redirect/main/ios/createproject.md

Launch Xcode and create a new iOS app project using SwiftUI for the interface and Swift for the language.

> > **Note:** The sample code uses the async/await feature of iOS 15 and later. Replacing those with calls that use completion handlers adds compatibility for iOS 14 and later.

Reviewer: ```suggestion
> **Note:** The sample code uses the async/await feature of iOS 15 and later. Replacing those with calls that use completion handlers adds compatibility for iOS 14 and earlier.
```

Author: Nice catch, it will be in the next commit.

> > **Note:** The sample code uses features of SwiftUI available in iOS 14, and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.

Author: In this case I didn't use code font because that's the way Apple presents the term on their developer website. At the moment I haven't made this change.

> > **Note:** The sample code uses features of SwiftUI available in iOS 14, and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.

Reviewer 2: Makes good sense. Thanks!

> > **Note:** The sample code uses features of SwiftUI available in iOS 14, and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.

Author: Thanks! It's in the next commit.

#### Add the UI

> The functionality of the app is implemented in the main SwiftUI view, including view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then using Combine, bindings, or `@Environment` for sharing state.

Author: Much nicer wording. I did modify it a little as I realized I forgot to refer to the file containing the view, and to tighten it up a bit, how about...

```suggestion
The functionality of the app is implemented in `ContentView.swift`, the main SwiftUI view. This includes view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then sharing state using Combine, bindings, or `@Environment`.
```

> The functionality of the app is implemented in the main SwiftUI view, including view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then using Combine, bindings, or `@Environment` for sharing state.

Reviewer 2: Works for me.

> The functionality of the app is implemented in `ContentView.swift`, the main SwiftUI view. This includes view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then sharing state using Combine, bindings, or `@Environment`.

Author: It's an Apple trademark that refers to a well known technology. I chose not to link to Apple developer documentation because it's both off site, and the possibility of links changing.

So, yes, it requires a capital and the expectation is that iOS developers will know what it means. Unless you have an objection, I'm leaving it as is.

> The functionality of the app is implemented in `ContentView.swift`, the main SwiftUI view. This includes view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then sharing state using Combine, bindings, or `@Environment`.

Reviewer 3: OK then that's jsut me not knowing Apple tech. This is fine.

> The text views in the app display its current state and any optional information, such as the access token. The available buttons depend on the sign-in state of the user. The top-level `ZStack` contains a full-screen view that prevents the user tapping the UI the app is performing certain asynchronous operations, such as reading the user information. A production app may also perform asynchronous operations that do require showing a busy indicator but don't require blocking the UI.

Reviewer: ```suggestion
The text views in the app display its current state and any optional information, such as the access token. The available buttons depend on the sign-in state of the user. The top-level `ZStack` contains a full-screen view that prevents the user interacting with the UI while the app is performing certain asynchronous operations, such as reading the user information. A production app may also perform asynchronous operations that do require showing a busy indicator but don't require blocking the UI.
```

Author: Nice catch. It will be in the next commit

> The text views in the app display its current state and any optional information, such as the access token. The available buttons depend on the sign-in state of the user. The top-level `ZStack` contains a full-screen view that prevents the user interacting with the UI while the app is performing certain asynchronous operations, such as reading the user information. A production app may also perform asynchronous operations that do require showing a busy indicator but don't require blocking the UI.

Author: Done. In the next commit.

The following image shows the completed app in both the sign-in and signed-on states:

<div class="full border">

![Two iPhone screenshots that show the signed-out and signed-in UI.](/img/sign-users-in/redirect-authentication/ios/app-ui-ios.png)

</div>

<!--
Source image(s): https://www.figma.com/file/i3huE0gEoISu2evquOq5yJ/app-ui-redirect-ios?node-id=0%3A1
There's only one board. The group for the image is: "Side-by-side screenshots"
-->

> 

Author: Yep, I know about the lack of space. I asked Chris and he said he'd add a margin to the styles... i.e., don't do anything about it myself either with a `<br/>` or `div style`.

> Open `ContentView.swift` and replace it with the following code:

Author: Nice catch. I slightly modified it as I realized the answer is "neither". It's only replacing any import statements and the ContentView struct. Replacing the contents of the file would also remove the preview struct as well. How's this:

```suggestion
Open `ContentView.swift` and use the following code to replace the `import` statements and the `ContentView` struct:```

> Open `ContentView.swift` and use the following code to replace the `import` statements and the `ContentView` struct:

Author: It's more than the struct, it's also the import statements. This is important because a SwiftUI file has at least 3 parts that should be there: the framework import statements, the struct defining the view, the preview struct. With that last item there's no dynamic preview in the development environment. I didn't want to add those lines of code as they're not relevant to the task at hand.

That's why it's replacing the import statements and view struct. If you'd still like to change the wording, do you have a suggestion? Or is OK as is?

> Open `ContentView.swift` and use the following code to replace the `import` statements and the `ContentView` struct:

Reviewer 3: I don't think switching the text around as suggested makes it any less obvious that both are being replaced because otherwise you appear to be omitting what to repalce the import statements with. I don't see anything about using an ampersand in the style guide, so 

> replace the `import` statements & the `ContentView` struct with the following code:

joins the two together more obviously. If that still doesn't work for you, leave as is.

> Open `ContentView.swift` and use the following code to replace the `import` statements and the `ContentView` struct:

Author: OK... this is what's in the next commit:

> Open `ContentView.swift` and replace the `import` statements and the `ContentView` struct with the following code:

> Open `ContentView.swift` and use the following code to replace the `import` statements and the `ContentView` struct:

Reviewer 3: That's lovely. Thank you.

```swift
import SwiftUI
import AuthFoundation
import WebAuthenticationUI

struct ContentView: View {

>    // Used for displaying error messages.

Reviewer 2: Refresh my memory re: comments in code - if it's a fragment, do we add a period?
```suggestion
   // Used for displaying error messages.
```

Author: I seem to remember from a previous article reviewed by Susan that the answer is "yes" and I can't find it in any of the various and sundry Okta style guides. I'm going to leave this for now. If there's a style issue, Susan will catch it in that final pass.

>    // Used for displaying error messages.

Author: Thanks. In the next commit.

   @State var showingError: Bool = false
   @State var errorMessage: String? = nil
   @State var errorTitle: String? = nil

>    // Text shown in the status area.

Author: Thanks. In the next commit.

               } else {
                  VStack {
                     HStack {
                        Button("Sign Out") {
                           signOut()
                        }
                        .padding()
                     }
                     HStack {
                        Button("Show User Info") {
                           showUserInfo()
                        }
                        .padding()
                        Button("Show Token") {
                           showTokenInfo()
                        }
                        .padding()
                     }
                  }
               }
            }
            Divider()
            Text("Status")
               .fontWeight(.medium)
               .foregroundColor(Color.gray)
            Text(statusText ?? "Unknown status.")
            Divider()
            Text("Info")
               .fontWeight(.medium)
               .foregroundColor(Color.gray)
            ScrollView(showsIndicators: true) {
               Text(infoText ?? "")
            }
         }
         if busy {
            Color.gray
               .opacity(0.5)
            ProgressView("Waiting")
        }
      }

>       .alert(isPresented: $showingError) {

Reviewer 4: Is this a typo? It looks like `$showingError` wouldn't compile.

         Alert(title: Text(errorTitle ?? "Unknown Error"),
         message: Text(errorMessage ?? "An unknow error occured."))
      }
   }

   func signIn() async {
      print("Sign in tapped")
   }

   func signOut() async {
      print("Sign out tapped")
   }

   func showTokenInfo() {
      print("Show token info tapped")
   }

   func showUserInfo() async {
      print("Show user info tapped")
   }

   // Display and error message using an Error object.
   func showError(title: String, error: Error) {
      showError(title: title, message: error.localizedDescription)
   }

   // Display an error message using a provided string.
   func showError(title: String , message: String) {
      errorTitle = title
      errorMessage = message
      showingError = true
   }

>    // A utility function to update the state of all visual elements.

Reviewer 5: ```suggestion
   // A utility function that updates the state of all visual elements.
```

Author: Thanks. In the next commit.

   func updateStatus(_ statusText: String, infoText: String, signedInStatus: Bool) {
      self.statusText = statusText
      self.infoText = infoText
      isSignedIn = signedInStatus
   }
}
```

> The last three functions are utilities. The two variants of `showError` set the state variables for the title and text of an alert, then set `showingError` which triggers the `.alert` view modifier to present the alert.

Author: Change is in next commit.

Launch Xcode and create a new iOS app project using SwiftUI for the interface and Swift for the language.

> > **Note:** The sample code uses SwiftUI features that are available in iOS 14 and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.

Reviewer: This makes it sound like only iOS 14+ is supported, when in fact older versions are supported.

```suggestion
> **Note:** The sample code uses SwiftUI features that are available in iOS 14 and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and earlier.
```

Author: I think we had this same exchange before. Although the SDK works with earlier versions, this sample won't. It uses features of SwiftUI that are iOS 14 or later, which is why the initial part of the sentence is there. If that's not clear then I need to revise the second sentence... how about:

>  Use the completion handler version of the async/await functions to enable the sample to run on iOS 14 and later.

Reviewer: I like that better, it makes it clearer that it's talking about the sample app itself.

> > **Note:** The sample code uses SwiftUI features that are available in iOS 14, and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.

Reviewer 5: ```suggestion
> **Note:** The sample code uses SwiftUI features that are available in iOS 14 and the async/await feature of iOS 15. Use the completion handler version of the async/await functions to add compatibility with iOS 14 and later.
```

#### Add the UI

The functionality of the app is implemented in `ContentView.swift`, the main SwiftUI view. This includes view state, view model updates, and data model. For a production app, consider partitioning the view and data models into singletons, and then sharing state using Combine, bindings, or `@Environment`.

> The app contains three areas: the top area contains buttons for signing in and out, and for viewing information, the middle area shows the current app state, and the lower area displays information text, such as the token. The available buttons depend on the sign-in state of the user. The top-level `ZStack` contains a full-screen view that prevents the user from interacting with the UI while the app is performing certain asynchronous operations, such as reading the user information. A production app may also perform asynchronous operations that do require showing a busy indicator but don't require blocking the UI.

Reviewer 5: ```suggestion
The app contains three areas: the top area contains buttons for signing in and out and for viewing information. The middle area shows the current app state, and the lower area displays information text, such as the token. The available buttons depend on the sign-in state of the user. The top-level `ZStack` contains a full-screen view that prevents the user from interacting with the UI while the app is performing certain asynchronous operations, such as reading the user information. A production app may also perform asynchronous operations that do require showing a busy indicator but don't require blocking the UI.
```

        }
      }
      .alert(isPresented: $showingError) {
         Alert(title: Text(errorTitle ?? "Unknown Error"),
         message: Text(errorMessage ?? "An unknow error occured."))
      }
   }

   func signIn() {
      infoText = "Sign in tapped"
   }

   func signOut() {
      infoText = "Sign out tapped"
   }

   func showUserInfo() {
      infoText =  "Show user info tapped"
   }

   func showTokenInfo() {
      infoText = "Show token info tapped"
   }

   func refreshToken() {
      infoText = "Refresh token tapped"
   }

   // Display and error message using an Error object.
   func showError(title: String, error: Error) {
      showError(title: title, message: error.localizedDescription)
   }

   // Display an error message using a provided string.
   func showError(title: String , message: String) {
      errorTitle = title
      errorMessage = message
      showingError = true
   }

   // A utility function that updates the state of all visual elements.
   func updateStatus(_ statusText: String, infoText: String, signedInStatus: Bool) {
      self.statusText = statusText
      self.infoText = infoText
      isSignedIn = signedInStatus
   }
}
```

> The last three functions are utilities. The two overloads of `showError` set the state variables for the title and text of an alert, then set `showingError` which triggers the `.alert` view modifier to present the alert.

Reviewer 5: ```suggestion
The last three functions are utilities. The two overloads of `showError` set the state variables for the title and text of an alert, then set `showingError` that triggers the `.alert` view modifier to present the alert.
```
